ksqldb-engine/src/main/java/io/confluent/ksql/engine/TransientQueryCleanupService.java

```diff
@@ -0,0 +1,312 @@
+        final Callable<Boolean> task = stateDirsCleanupTasks.take();
+
+        try {
+          final boolean cleanedUp = task.call();
+          if (!cleanedUp) {
+            stateDirsCleanupTasks.add(task);
+          }
+        } catch (Exception e) {
+          stateDirsCleanupTasks.add(task);
+        }
+      }
+    } catch (final InterruptedException e) {
+      // gracefully exit if this method was interrupted and reset
+      // the interrupt flag
+      Thread.currentThread().interrupt();
+    }
+  }
+
+  @SuppressFBWarnings(value = "EI_EXPOSE_REP2")
+  public void setLocalCommandsQueryAppIds(final Set<String> ids) {
+    this.localCommandsQueryAppIds = ids;
+    this.isLocalCommandsInitialized = true;
+  }
+
+  public void setQueryRegistry(final QueryRegistry queryRegistry) {
+    this.queryRegistry = queryRegistry;
+  }
+
+  public void addTopicCleanupTask(final TransientQueryTopicCleanupTask task) {
+    topicsCleanupTasks.add(task);
+  }
+
+  public void addStateCleanupTask(final TransientQueryStateCleanupTask task) {
+    stateDirsCleanupTasks.add(task);
+  }
+
+  private List<TransientQueryStateCleanupTask> findPossiblyLeakedStateDirs() {
+    final String stateDir = this.stateDir;
+    final File folder = new File(stateDir);
+    final File[] listOfFiles = folder.listFiles();
+
+    if (listOfFiles == null) {
+      return Collections.emptyList();
+    }
+
+    final List<TransientQueryStateCleanupTask> leakedStates
+            = new LinkedList<>();
+
+    for (File f: listOfFiles) {
+      final String fileName = f.getName();
+      final Matcher filenameMatcher = transientPattern.matcher(fileName);
+      if (filenameMatcher.find() && !isCorrespondingQueryRunning(f.getName())) {
+        LOG.info("{} seems to be a leaked state directory. Adding it to the cleanup queue.",
+                f.getName());
+
+        leakedStates.add(new TransientQueryStateCleanupTask(
+                f.getName(),
+                stateDir)
+        );
+      }
+    }
+
+    return leakedStates;
+  }
+
+  private List<TransientQueryTopicCleanupTask> findPossiblyLeakedTransientTopics() {
+    final List<TransientQueryTopicCleanupTask> leakedQueries
+            = new LinkedList<>();
+
+    for (String topic : this.serviceContext
+            .getTopicClient()
+            .listTopicNames()) {
+      final Matcher topicNameMatcher = transientPattern.matcher(topic);
+      if (topicNameMatcher.find() && !isCorrespondingQueryRunning(topic)) {
+        LOG.info("{} topic seems to have leaked. Adding it to the cleanup queue.", topic);
+        final Matcher topicPrefixMatcher = LEAKED_TOPIC_PREFIX_PATTERN.matcher(topic);
+        if (topicPrefixMatcher.find()) {
+          final String leakedTopicPrefix = topicPrefixMatcher.group();
+          leakedQueries.add(new TransientQueryTopicCleanupTask(
+                  serviceContext,
+                  leakedTopicPrefix
+          ));
+        }
+      }
+    }
```

Comment:
Quick check here: is this only going to match topics for transient queries created by _this_ node, or will we also see topics created by transient queries on other nodes? In the latter case, it seems like we'd delete the topic for running queries on other hosts, right?